Add Blueprint Editor

jpitlor opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
I wanted to add an editor for Blueprints, similar to Automations and Scripts. Essentially, this creates a page that has an editor for `blueprintObject.metadata.input`, then embeds the existing automation/script editor.

 There was a [WTH forum post](https://community.home-assistant.io/t/wth-cant-we-create-or-edit-blueprints-with-the-ui/469776/5) asking for this. I'm not sure if there were any formal feature requests/roadmap plans, but truthfully this is mostly inspired by my own desire for this feature to exist.

## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->
I don't think any particular configuration is needed to test this, especially since HA comes with a couple blueprints out of the box.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
- This PR fixes or closes issue: N/A
- This PR is related to issue or discussion: https://community.home-assistant.io/t/wth-cant-we-create-or-edit-blueprints-with-the-ui/469776/5
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/39316
- Link to core pull request: https://github.com/home-assistant/core/pull/145863

Additionally, I have concerns about the size of this MR. I think it's quite large. If it should be broken up, I'd love to do that. I just need guidance on the best strategy. However, there are parts of this feature I left out explicitly because of the size:
- a `/show` route 
- Adding support for `!input`s in visual selectors
- ~~[Selector Configuration](https://www.home-assistant.io/docs/blueprint/selectors/)~~ Implemented, thanks to feedback from karwosts!
- Validation/testing (of the blueprints, not the code)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works. _I did not do this because it doesn't seem like there's a way to test UI components. At least, I didn't see anything in the `test` directory. Let me know if there's something I can look at!_

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

## Screenshots

<img width="1094" height="758" alt="Screenshot 2026-02-05 at 10 34 15 PM" src="https://github.com/user-attachments/assets/2fc2665e-8ed0-419c-803c-66f331dc748c" />

<img width="2258" height="1504" alt="Firefox_Screenshot_2026-02-06T03-19-40 637Z" src="https://github.com/user-attachments/assets/a20683ff-b416-463b-be10-166421312de7" />

jpitlor committed:
Initial implementation of blueprint editor

jpitlor committed:
Some bug fixes

jpitlor committed:
Remove unneeded changes

jpitlor committed:
Add blueprint input sections

jpitlor committed:
Remove remaining errors

jpitlor committed:
Remove remaining errors

jpitlor committed:
Remove comments and TODOs

jpitlor committed:
Merge `dev`

jpitlor committed:
Remove unused imports

jpitlor committed:
Remove invalid selector types

jpitlor committed:
Fix incorrect redirect after saving blueprint bug

jpitlor committed:
Add TODOs for inputs in YAML

jpitlor committed:
Add support for arbitrary inputs in YAML editors for blueprints

jpitlor committed:
Remove unused translations for blueprints

jpitlor committed:
Fix YAML related error

jpitlor committed:
Fix Save Blueprint API call

jpitlor committed:
Merge branch 'dev' into blueprint-editor

karwosts commented on src/panels/config/blueprint/input/types/ha-blueprint-input-input.ts:
```diff
@@ -0,0 +1,332 @@
+          name: "default",
+          selector: defaultSelector,
+        },
+        // TODO: Sub-selector settings (https://www.home-assistant.io/docs/blueprint/selectors/)
```
We have a `ha-selector-selector` already that allows for picking a selector type and configuring its options and defaults. Kind of feels like that's what you're trying to do here, maybe you can leverage that instead?

It's currently used in the script editor for `Add Fields`.

jpitlor replied:
Yes, that was exactly what I wanted! Thanks for letting me know about that. Just pushed an update - that helped cut down on a lot of code!

jpitlor committed:
Utilize selector selector

jpitlor committed:
Merge remote-tracking branch 'origin/blueprint-editor' into blueprint-editor

NoRi2909 commented on src/translations/en.json:
```diff
@@ -4527,6 +4538,78 @@
+            "load_error_unknown": "Error loading blueprint ({err_no}).",
+            "move_up": "Move up",
+            "move_down": "Move down",
+            "edit_ui": "UI Editor",
+            "edit_yaml": "YAML Editor",
```
We need sentence-casing on these strings and a bunch more below.

```
"edit_ui": "UI editor",
"edit_yaml": "YAML editor",

"delete_confirm_title": "Delete input",

"change_id": "Change input ID",

"default": "Default value",

"add_description": "Add description",
"add_author": "Add author",
"add_minimum_version": "Add minimum version"

"label": "Minimum version",

```

I may have missed another one.

jpitlor replied:
will do! I'll double check all of the ones I added. Is the general guidance to use sentence casing on everything? Or are there any exceptions?

NoRi2909 replied:
Yes, Home Assistant uses sentence-casing in general. It helps translators like me to see what can be translated and what is a (capitalized) name or abbreviation that we should keep in translations.

jpitlor replied:
That makes sense! Thanks for pointing it out. I'll try to get to this some time today

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
Make strings sentence case

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
Fix lint errors

jpitlor committed:
Update src/translations/en.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

jpitlor commented on src/translations/en.json:
```diff
@@ -4545,6 +4556,78 @@
             "error_no_url": "Please enter the blueprint address.",
             "unsupported_blueprint": "This blueprint is not supported",
             "file_name": "Blueprint Path"
```
How convenient that Github lets me commit this straight from the web UI! Done!

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
Fix lint errors

jpitlor committed:
Merge branch 'dev' into blueprint-editor

silamon commented:
I've done a test run on this one and it work pretty well but I'm a bit thinking about how it will be used by home assistant users that don't manage the original blueprint.  

The definition behind blueprints is that people can share them and other people can import them to use them in their home assistant installation. If there are updates to the blueprint, you can redownload it from the location you've initially downloaded them. If the user thinks that the blueprint doesn't fit their use-case completely and wants to modify it, they must take control and get a regular automation or script they can modify.  

The blueprint editor introduced here, is just extending the regular blueprint area. The gap between people that develop blueprints and people that use blueprints is pretty close. People that don't manage the blueprint will just start editing the blueprint in the handy editor, having a modified version to their likings which will overwrite the original blueprint straightaway without warning, and they won't know anymore after a few months and they'll lose their changes when the blueprint will be updated by them. 

TLDR; I think this nice blueprint editor may be a better fit under the developer tools.

jpitlor commented:
Thanks for taking a look! Yeah that's completely fair - I didn't even realize you could update blueprints, so I agree that the developer tools is probably a better fit for this. I would still be satisfied - an editor in the developer tools is still better than no editor.

I can move it there, although I'm out of town this week, so it probably won't happen until some time next week

jpitlor commented:
Thanks for the patience - I'm just now getting time to make progress on this again

@silamon as I'm moving the editor to the developer tools, I'm trying to think of how it should be used and I'd love to have your input since you know how people use this feature better than I do.

I'm assuming a UI editor that spits out the corresponding YAML to create a blueprint and then copy it over to GitHub (or wherever people put their blueprints) has a high likelihood of being valuable, especially since it mimics the purpose of the template tab.

I'm also assuming the ability to add your own blueprints may be valuable, as maybe people want to create their own blueprint, but they don't want to go through the effort of hosting them somewhere. Does this break any part of the blueprint page if the blueprint doesn't have an associated URL for updates?

And then there's the ability to edit existing blueprints - should I just leave this part out, or should I put it in with some warning dialog to add friction and make sure people know what they're doing?

jpitlor committed:
Move blueprint editor to new developer tools tab

silamon commented:
That need some thinking. Since I'm just a voluntary reviewer I may wait for feedback from the Nabu Casa team.

jpitlor committed:
Move add metadata to the editor

jpitlor committed:
Merge branch 'refs/heads/dev' into blueprint-editor

# Conflicts:
#	src/panels/config/automation/action/ha-automation-action-row.ts
#	src/panels/config/automation/blueprint-automation-editor.ts
#	src/panels/config/automation/condition/ha-automation-condition-editor.ts
#	src/panels/config/automation/manual-automation-editor.ts
#	src/panels/config/automation/paste-replace-dialog/dialog-paste-replace.ts
#	src/panels/config/automation/trigger/ha-automation-trigger-row.ts
#	src/panels/config/blueprint/ha-blueprint-overview.ts
#	src/panels/config/blueprint/import-blueprint-dialog/dialog-import-blueprint.ts
#	src/panels/config/script/blueprint-script-editor.ts
#	src/panels/config/script/ha-script-editor.ts
#	src/panels/config/script/ha-script-field-row.ts
#	src/panels/developer-tools/ha-panel-developer-tools.ts

jpitlor committed:
Made some initial progress on moving the blueprint editor to developer tools.

Still need to work on saving (and all the edge cases surrounding it)

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
More progress on Blueprint Editor

jpitlor committed:
Merge branch 'refs/heads/dev' into blueprint-editor

# Conflicts:
#	src/panels/config/automation/action/ha-automation-action-row.ts

jpitlor committed:
Finish moving blueprint editor to developer tools

jpitlor committed:
Merge branch 'dev' into blueprint-editor

# Conflicts:
#	src/panels/config/blueprint/ha-config-blueprint.ts

jpitlor committed:
Remove old changes that are no longer needed

jpitlor committed:
Fix bug where blueprint never loads

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor commented:
yeesh! sorry gang, that took way longer than I expected. I suppose these things happen with open source development. Thanks again for your patience, this should be ready for review again.

jpitlor committed:
Merge branch 'dev' into blueprint-editor

silamon commented:
Things will move faster now that core is already approved, I'll give it a review this weekend. Feel free to resolve the linting issues in the meantime.

jpitlor committed:
Fix lint error

jpitlor committed:
Merge remote-tracking branch 'origin/blueprint-editor' into blueprint-editor

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
Fix lint errors

jpitlor committed:
Merge remote-tracking branch 'origin/blueprint-editor' into blueprint-editor

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor commented:
I totally missed the email that the lint was failing, so sorry about that!

silamon commented on src/panels/config/automation/condition/ha-automation-condition-editor.ts:
```diff
@@ -2,6 +2,7 @@ import { css, html, LitElement, nothing } from "lit";
 import { customElement, property, query } from "lit/decorators";
 import { classMap } from "lit/directives/class-map";
 import memoizeOne from "memoize-one";
+import { DEFAULT_SCHEMA, Schema } from "js-yaml";
```
```suggestion
import { DEFAULT_SCHEMA} from "js-yaml";
import type { Schema } from "js-yaml";
```

silamon commented on src/panels/config/automation/condition/ha-automation-condition-editor.ts:
```diff
@@ -13,6 +14,8 @@ import type { HomeAssistant } from "../../../../types";
 import "../ha-automation-editor-warning";
 import { editorStyles, indentStyle } from "../styles";
 import type { ConditionElement } from "./ha-automation-condition-row";
+import { consume } from "@lit/context";
```
Place this right after the other lit imports.

silamon commented on src/panels/config/automation/trigger/ha-automation-trigger-row.ts:
```diff
@@ -15,6 +15,7 @@ import {
   mdiStopCircleOutline,
 } from "@mdi/js";
 import type { UnsubscribeFunc } from "home-assistant-js-websocket";
+import { DEFAULT_SCHEMA, Schema } from "js-yaml";
```
```suggestion
import { dump, DEFAULT_SCHEMA} from "js-yaml";
import type { Schema } from "js-yaml";
```

silamon commented on src/panels/config/automation/trigger/ha-automation-trigger-row.ts:
```diff
@@ -15,6 +15,7 @@ import {
 } from "@mdi/js";
 import type { UnsubscribeFunc } from "home-assistant-js-websocket";
+import { DEFAULT_SCHEMA, Schema } from "js-yaml";
 import { dump } from "js-yaml";
```
```suggestion
```

silamon reviewed (commented):
Linting issues

silamon reviewed (changes requested):
I've left some fine tunings in the comments above. If you need help on some of the topics, don't hesitate to ask.

jpitlor committed:
Update src/panels/config/automation/condition/ha-automation-condition-editor.ts

Co-authored-by: Simon Lamon <32477463+silamon@users.noreply.github.com>

jpitlor committed:
Update src/panels/config/automation/trigger/ha-automation-trigger-row.ts

Co-authored-by: Simon Lamon <32477463+silamon@users.noreply.github.com>

jpitlor committed:
Update src/panels/config/automation/trigger/ha-automation-trigger-row.ts

Co-authored-by: Simon Lamon <32477463+silamon@users.noreply.github.com>

jpitlor committed:
Blueprint Editor:
- Make save button FAB
- Make reset button part of blueprint card
- Remove default values in new blueprint
- UX/styling improvements

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
Blueprint Editor:
- Make save button FAB
- Make reset button part of blueprint card
- Remove default values in new blueprint
- UX/styling improvements

jpitlor committed:
Blueprint Editor: prettier format
- Make save button FAB
- Make reset button part of blueprint card
- Remove default values in new blueprint
- UX/styling improvements

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor commented:
@silamon just double checking - there's nothing you're waiting on me for, right? I'm happy to wait, I would just hate to not realize part of the workflow is waiting on me.

silamon committed:
Merge branch 'dev' into blueprint-editor

silamon commented:
The bottom-sheet (the thing that shows on the right or bottom when on mobile upon adding an action) is still a bit broken. Unfortunately, I haven't had time yet to see how it works in the automation/script editor to guide you further. Let me try to pull in the author of that component to guide you further. 

Others than that, the improvements you've made in the meantime are really appreciated! :)

jpitlor committed:
Merge branch 'home-assistant:dev' into blueprint-editor

jpitlor committed:
Blueprint Editor: correct some of the styles for the sidebar

jpitlor committed:
Blueprint Editor: fix initial padding for sidebar content and sidebar positioning

jpitlor committed:
Merge branch 'dev' into blueprint-editor

wendevlin commented on src/translations/en.json:
```diff
@@ -9489,6 +9489,100 @@
+              },
+              "minimum_version": {
+                "label": "Minimum version",
+                "placeholder": "2025.1.0"
```
```suggestion
```

seems unused

wendevlin commented on src/translations/en.json:
```diff
@@ -9489,6 +9489,100 @@
+              },
+              "author": {
+                "label": "Author",
+                "placeholder": "Linus Torvalds"
```
```suggestion
```

seems unused and we shouldn't use a real person here.

wendevlin commented on src/translations/en.json:
```diff
@@ -9489,6 +9489,100 @@
+                "add_description": "Add description",
+                "add_author": "Add author",
+                "add_minimum_version": "Add minimum version"
+              },
```
is this used? Please check all translations again if they are used.

wendevlin commented on src/data/blueprint.ts:
```diff
@@ -1,24 +1,83 @@
   description?: string;
   source_url?: string;
   author?: string;
+  homeassistant?: { min_version: string };
```
on the one hand it's minimum_version on the other `min_version` I assume `min_version` is the blueprint schema so keep that on both please.

wendevlin commented on src/panels/config/automation/action/types/ha-automation-action-event.ts:
```diff
@@ -22,6 +26,9 @@ export class HaEventAction extends LitElement implements ActionElement {
   @query("ha-yaml-editor", true) private _yamlEditor?: HaYamlEditor;
 
+  @consume({ context: yamlSchemaContext })
+  private _yamlSchema?: Schema;
```
As far as I understand this, it's never set. So we always use `DEFAULT_SCHEMA` right? `DEFAULT_SCHEMA` is set as default in yaml editor. Why do we add it here?

jpitlor replied:
Yeah, this does seem a little weird. I suspect I realized all of the YAML editors for each row would need the blueprint schema, but forgot about testing this use case before submitting the PR (I don't often use the YAML editor, not that that's an excuse not to test)

I think given the number of files this would touch, it may make sense to remove these changes from this PR and add "make the YAML editors work" as part of the second PR that makes the visual selectors work too.

jpitlor replied:
But to be clear, the correct implementation is to keep this and also set the context to the BlueprintEditorSchema when a blueprint is being edited

wendevlin commented on src/panels/config/developer-tools/blueprints/input/types/ha-blueprint-input-input.ts:
```diff
@@ -0,0 +1,99 @@
+    const schema = this._schema(selector);
+
+    return html`
+      <ha-form
```
import is missing

wendevlin commented on src/panels/developer-tools/blueprints/input/types/ha-blueprint-input-input.ts:
```diff
@@ -0,0 +1,99 @@
+    ha-select,
+    ha-selector {
+      display: block;
+      margin-bottom: 24px;
```
```suggestion
      margin-bottom: var(--ha-space-6);
```

wendevlin commented on src/panels/developer-tools/blueprints/pick-blueprint-dialog/dialog-pick-blueprint.ts:
```diff
@@ -0,0 +1,211 @@
+      Blueprint,
+    ][];
+    return html`
+      <mwc-list
```
please use `ha-list` or even better `ha-md-list`

wendevlin commented on src/panels/config/developer-tools/blueprints/blueprint-metadata-editor.ts:
```diff
@@ -0,0 +1,111 @@
+          </h2>
+        </div>
+        <div class="content">
+          <ha-form
```
import missing

wendevlin commented on src/panels/developer-tools/blueprints/blueprint-metadata-editor.ts:
```diff
@@ -0,0 +1,111 @@
+        .container {
+          margin: 0 auto;
+          max-width: 1040px;
+          padding: 8px 20px 0;
```
```suggestion
          padding: var(--ha-space-2) var(--ha-space-5) 0;
```

wendevlin commented on src/panels/developer-tools/blueprints/blueprint-metadata-editor.ts:
```diff
@@ -0,0 +1,111 @@
+        }
+
+        .header {
+          margin-top: 16px;
```
```suggestion
          margin-top: var(--ha-space-4);
```

wendevlin commented on src/panels/developer-tools/blueprints/blueprint-metadata-editor.ts:
```diff
@@ -0,0 +1,111 @@
+        }
+
+        .header:first-child {
+          margin-top: -16px;
```
```suggestion
          margin-top: calc(var(--ha-space-4) * -1);
```

wendevlin commented on src/panels/developer-tools/blueprints/blueprint-metadata-editor.ts:
```diff
@@ -0,0 +1,111 @@
+        .header .name {
+          font-weight: var(--ha-font-weight-normal);
+          flex: 1;
+          margin-bottom: 8px;
```
```suggestion
          margin-bottom: var(--ha-space-2);
```

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+    const [automation, script] = await Promise.all([
+      fetchBlueprints(this.hass, "automation"),
+      fetchBlueprints(this.hass, "script"),
+    ]);
```
add a try catch and handle possible errors

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+      };
+    }
+    this._dirty = false;
+  }
```
seems to be duplicate code. You could just make one func out of it and provide a param for diry

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+        text: this.hass.localize(
+          "ui.panel.developer-tools.tabs.blueprints.editor.abandon_changes_text"
+        ),
+      });
```
```suggestion
      const shouldContinue = await showConfirmationDialog(this, {
        title: this.hass.localize(
          "ui.panel.developer-tools.tabs.blueprints.editor.abandon_changes_title"
        ),
        text: this.hass.localize(
          "ui.panel.developer-tools.tabs.blueprints.editor.abandon_changes_text"
        ),
        destructive: true
      });
```

make the button red maybe also edit `confirmText` to something more descriptive then OK

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+      dump(this._selectedBlueprint),
+      this._selectedBlueprint.metadata.source_url,
+      true
+    );
```
add a try catch block. If you have invalid yaml core will throw an error and you don't inform the user.

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+
+    const blueprints = Object.values(this._blueprints)
+      .flatMap((b) => Object.values(b))
+      .filter((b) => !("error" in b)) as Blueprint[];
```
this could be memoized

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+      .flatMap((b) => Object.values(b))
+      .filter((b) => !("error" in b)) as Blueprint[];
+    const blueprintMetadata =
+      !this._selectedBlueprint || "error" in this._selectedBlueprint
```
You use this pretty often. We could make a func out of it, for better readability

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+          .autoUpdate=${true}
+          .readOnly=${true}
+        >
+        </ha-yaml-editor>
```
This should be switchable like in the automation editor, either UI or yaml editor. and then it should be editable.

jpitlor replied:
I have UX concerns about this, and I just want to make sure you're sure about it. 

If this were in the blueprints page, I would agree 100% (and this is actually how it was before it moved to the developer tools). If someone is here to make a blueprint to copy over to GitHub, we can be pretty confident they'll want both the UI and the YAML. 

Is this too much friction? Also, the YAML switches are usually in overflow menus, so it might not be obvious how to get the YAML (although we can probably assume the developers know where the YAML switches are, so this may not be relevant).

If you're confident this is how you want it, I can change it, I just want to make sure.

jpitlor replied:
Revisiting this - I'm seeing some _massive_ performance problems when the blueprint is large and the firefox profiler is saying it's caused by constant reparsing of the yaml in `ha-yaml-editor`.

Looking at the `codemirror` documentation, I do see a `view.state.update` method that is more meant for an application like this, but considering this is the only place this is a problem, and considering you want to potentially move this feature out of the developer tools, I think it actually makes more sense to implement this.

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+      manualEditorStyles,
+      css`
+        :host {
+          --safe-area-inset-top: 52.4px;
```
Whats that? Why would we set the safe area manually?

jpitlor replied:
So this was my way of getting the sidebar on the automation editor positioned correctly. It seemed dirty, and I'd love to know if there's a better way of doing it. 

As I understand it, the sidebar is positioned to `top: calc(safe-area-inset-top + some padding)`, and the default safe area is the height of the header. The problem is that in the developer tools specifically, the tabs make the header taller, so without this, the sidebar extends into the header.

Again, if there's a better way to get this to work, I'd love to change this

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+        }
+
+        ha-card {
+          padding: 8px;
```
```suggestion
          margin: var(--ha-space-4);
          gap: var(--ha-space-4);
          display: flex;
          flex-direction: column;
        }

        ha-card {
          padding: var(--ha-space-2);
```

wendevlin commented on src/panels/developer-tools/blueprints/developer-tools-blueprints.ts:
```diff
@@ -0,0 +1,431 @@
+        </ha-yaml-editor>
+      </div>
+
+      <ha-fab
```
import missing

wendevlin commented on src/panels/developer-tools/blueprints/ha-blueprint-automation-editor.ts:
```diff
@@ -0,0 +1,74 @@
+    conditions: [],
+    actions: [],
+  };
+
```
you can move this to `data/blueprint.ts`

wendevlin commented on src/panels/developer-tools/blueprints/ha-blueprint-generic-editor.ts:
```diff
@@ -0,0 +1,389 @@
+    // The HASS object is a prerequisite to load blueprints
+    if (!this.hass) {
+      return;
+    }
```
hass is mandatory in the props so this check can be avoided

wendevlin commented on src/panels/developer-tools/blueprints/input/ha-blueprint-input.ts:
```diff
@@ -0,0 +1,243 @@
+  protected render() {
+    if (!Array.isArray(this.inputs)) {
+      return nothing;
+    }
```
What about showing a yaml editor if it's not supported by UI?

jpitlor replied:
I'm actually going to take this out - looking back, `this.inputs` is calculated as `Object.entries(...)`, which always either returns an array or crashes (if the parameter is null)

wendevlin commented on src/panels/developer-tools/blueprints/input/ha-blueprint-input.ts:
```diff
@@ -0,0 +1,243 @@
+      flex-direction: column;
+      padding: 16px;
+      margin: -16px;
+      gap: 16px;
```
```suggestion
      padding: var(--ha-space-4);
      margin: calc(var(--ha-space-4) * -1);
      gap: var(--ha-space-4);
```

wendevlin commented on src/panels/developer-tools/blueprints/input/ha-blueprint-input-row.ts:
```diff
@@ -0,0 +1,451 @@
+
+    return html`
+      <ha-card outlined>
+        <ha-expansion-panel leftChevron>
```
You mix up the UI elements. Why is this not a `ha-automation-row`? It should work the same as triggers/conditions and actions. To have the same behavior when clicking on a row.

jpitlor replied:
Good point - I didn't even realize the script editor uses that element too! I'll reuse that element

wendevlin commented on src/panels/developer-tools/blueprints/input/ha-blueprint-input-row.ts:
```diff
@@ -0,0 +1,451 @@
+          pointer-events: none;
+        }
+        ha-expansion-panel {
+          --expansion-panel-summary-padding: 0 0 0 8px;
```
```suggestion
          --expansion-panel-summary-padding: 0 0 0 var(--ha-space-2);
```

wendevlin commented on src/panels/developer-tools/blueprints/input/ha-blueprint-input-row.ts:
```diff
@@ -0,0 +1,451 @@
+          background: var(--divider-color, #e0e0e0);
+          text-align: center;
+          border-top-right-radius: var(--ha-card-border-radius, 12px);
+          border-top-left-radius: var(--ha-card-border-radius, 12px);
```
```suggestion
            margin-right: var(--ha-space-2);
            margin-inline-end: var(--ha-space-2);
            margin-inline-start: initial;
          }
        }
        .card-content {
          padding: var(--ha-space-4);
        }
        .disabled-bar {
          background: var(--divider-color, #e0e0e0);
          text-align: center;
          border-top-right-radius: var(--ha-card-border-radius, var(--ha-border-radius-lg));
          border-top-left-radius: var(--ha-card-border-radius, var(--ha-border-radius-lg));
```

wendevlin commented on src/panels/developer-tools/blueprints/input/ha-blueprint-input-row.ts:
```diff
@@ -0,0 +1,451 @@
+          left: 0px;
+          text-transform: uppercase;
+          font-weight: bold;
+          font-size: 14px;
```
```suggestion
          font-size: var(--ha-font-size-m);
```

wendevlin commented on src/panels/developer-tools/blueprints/input/ha-blueprint-input-row.ts:
```diff
@@ -0,0 +1,451 @@
+          transition: max-height 0.3s;
+          text-align: center;
+          border-top-right-radius: var(--ha-card-border-radius, 12px);
+          border-top-left-radius: var(--ha-card-border-radius, 12px);
```
```suggestion
          border-top-right-radius: var(--ha-card-border-radius, var(--ha-border-radius-lg));
          border-top-left-radius: var(--ha-card-border-radius, var(--ha-border-radius-lg));
```

wendevlin reviewed (changes requested):
Hi @jpitlor,

This is an amazing addition. 
I added some comments.

We are going to discuss this with our UX team, we are not sure if dev tools is the right place for it.
I think in an first iteration it fits, but when it improves in the future it will fit more in the blueprints config.

Nice to have:
- auto suggest blueprint path for new blueprint (from name)

For now it just adds a UI for input values, but the values itself are not supported by the automation UI. This is not an easy part, do you think, you can add this in the future?

Thanks Wendelin

wendevlin commented:
> > We are going to discuss this with our UX team, we are not sure if dev tools is the right place for it. I think in an first iteration it fits, but when it improves in the future it will fit more in the blueprints config.
> 
> It was first in the blueprints config, but then I suggested to move it over to the dev tools. The reasoning behind is that this is a developer tool to get a clean UI to edit blueprints if you're developing them. It's in no way a tool for end users to modify blueprints to their needs. If we move it back to the blueprint config, we need to be able to distinct that clearly.

Yeah but when it works well it's fine to use as enduser and gives you the opportunity to create your own blueprints easily. I also tend to keep it in the dev tools but add a note that it may move in future.

jpitlor committed:
Apply suggestions from code review

Co-authored-by: Wendelin <12148533+wendevlin@users.noreply.github.com>

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
Update src/panels/developer-tools/blueprints/developer-tools-blueprints.ts

Co-authored-by: Wendelin <12148533+wendevlin@users.noreply.github.com>

jpitlor committed:
Blueprint Editor: implementing some PR feedback

jpitlor committed:
Blueprint Editor: remove unused strings

jpitlor committed:
Blueprint Editor: remove half-baked yaml schema context. The plan is to re-add it in a future PR

jpitlor committed:
Blueprint Editor: remove mwc-list uses

jpitlor committed:
Blueprint Editor: make a utility function for knowing if a blueprint is valid

jpitlor committed:
Blueprint Editor: put default blueprints in data file

jpitlor committed:
Blueprint Editor: Add error handling for saving blueprint

jpitlor committed:
Blueprint Editor: Refactor to put metadata editing in ha-blueprint-editor, refactor to inline logic from ha-blueprint-{automation,script}-editor

jpitlor committed:
Blueprint Editor: Fix several bugs, mostly relating to the sidebar

jpitlor committed:
Blueprint Editor: Fix bug preventing input changes

jpitlor committed:
Blueprint Editor: Implement input sections in sidebar

jpitlor committed:
Merge branch 'feature-blueprint-editor' into blueprint-editor

jpitlor committed:
Blueprint Editor: Fix input section breadcrumb bugs

jpitlor committed:
Blueprint Editor: Fix editing inputs inside of sections

jpitlor committed:
Blueprint Editor: Remove unused show ID action

jpitlor committed:
Blueprint Editor: Use input names where available instead of IDs on labels

jpitlor committed:
Blueprint Editor: Add yaml/ui switch, remove incorrect usages of blueprint/metadata keys

jpitlor committed:
Blueprint Editor: Revert unneeded changes

jpitlor committed:
Blueprint Editor: Fix cut/copy/duplicate for top level and nested inputs

jpitlor committed:
Blueprint Editor: Fix bug where changing section input list didn't render in sidebar

jpitlor committed:
Blueprint Editor: Fix bug where changing section input list inside of a section didn't render in sidebar

jpitlor committed:
Blueprint Editor: Add initial implementation for copy/paste

jpitlor committed:
Blueprint Editor: Fix paste

jpitlor committed:
Blueprint Editor: Fix regressions introduced earlier

silamon committed:
Fix devcontainer

silamon committed:
Merge remote-tracking branch 'upstream/dev' into blueprint-editor

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor commented:
I pulled the changes from @silamon's branch, and I don't have enough time right now to do in depth re-testing, but I ran the client, visited the tab, and added an input to a blueprint, and nothing crashed, so I'll just assume the rest of the merge went well.

@wendevlin I think we're good for another review

jpitlor committed:
Blueprint Editor: Fix linting

jpitlor committed:
Merge remote-tracking branch 'origin/blueprint-editor' into blueprint-editor

jpitlor committed:
Merge branch 'dev' into blueprint-editor

jpitlor committed:
Blueprint Editor: Fix linting

jpitlor committed:
Merge remote-tracking branch 'origin/blueprint-editor' into blueprint-editor

jpitlor committed:
Merge branch 'dev' into blueprint-editor

silamon commented:
So review and if good, we're going to merge to feature branch. Please note that I've changed default branch to dev to see conflicts.

wendevlin commented on src/panels/config/developer-tools/blueprints/new-input-dialog/dialog-new-input.ts:
```diff
@@ -0,0 +1,144 @@
+          .computeLabel=${this._computeLabel}
+          @value-changed=${this._valueChanged}
+        ></ha-form>
+        <ha-button
+          variant="neutral"
+          @click=${this.closeDialog}
+          dialogInitialFocus
+          slot="secondaryAction"
+        >
+          ${this.hass.localize("ui.common.cancel")}
+        </ha-button>
+        <ha-button
+          .disabled=${!this._data.id}
+          @click=${this._submit}
+          slot="primaryAction"
+        >
+          ${this.hass.localize("ui.common.save")}
+        </ha-button>
```
```suggestion
        <ha-dialog-footer slot="footer">
          <ha-button
            variant="neutral"
            @click=${this.closeDialog}
            dialogInitialFocus
            slot="secondaryAction"
          >
            ${this.hass.localize("ui.common.cancel")}
          </ha-button>
          <ha-button
            .disabled=${!this._data.id}
            @click=${this._submit}
            slot="primaryAction"
          >
            ${this.hass.localize("ui.common.save")}
          </ha-button>
        </ha-dialog-footer>
```

wendevlin reviewed (commented):
Hey @jpitlor,

Yeah it looks much better, but I am not sure if we should continue.
I really appreciate your work here and that you keep pushing. Also the idea is really nice and will give us the possibility to create blueprints without knowing the yaml syntax.

But just with testing the UI I had this issues:
- Add new input dialog -> submit button missing
  - this is just a merge issue
- loading indicator missing when open a blueprint
  - loading a blueprint can take a while and we give 0 feedback
- bigger blueprints make the whole UI unsresponsive on a performant PC
- input section:  childs should be rows on the mainflow instead of the sidebar
- input sidebar: There seems to be a big issue how changes are handled. Sometimes it's not changing sometimes it changes back on the next change.
- path suggestion → use slugify
- blueprint should be selectable via the url, so a refresh of the editor should be possible

I think this is one of the most complex topic I can think of and it should be somehow maintainable in the future.

My suggestion would be to close this PR and you could use this version to develop a custom panel for it? Not sure how this work with your core changes. 

Or we keep working on it in 2 ways:

  1. I plan to refactor the whole automations code. Because we have a lot of code duplication there, what makes it complex to edit and the way we handle changes to and from the sidebar is really complex. I have the idea to maybe switch to a system that uses the lit context approach. 
  2. We work on a concept and mockup of this Blueprint editor in every detail together. To have at least a full idea how it should look and behave in the end.

When we have both restart the editor with a solid base we can reuse from the automations and a clear plan what to do.

Unfortunately I have to say you that this is not our nr. 1 prio this year, it won't be on our roadmap. Lets keep pushing but it can happen that I have to do other things with more priority.

I hope you are not too sad with my suggestion, I totally understand to you invested a lot of time and I am still in favor of this editor, I just think it needs to be rock solid from the start up and therefore we need a solid base from the automations.

Please feel free to discuss with us about it.
